iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl

```diff
@@ -65,74 +92,59 @@ inline bool SpscSofi<ValueType, CapacityValue>::setCapacity(const uint64_t newSi
 template <class ValueType, uint64_t CapacityValue>
 inline bool SpscSofi<ValueType, CapacityValue>::empty() const noexcept
 {
-    uint64_t currentReadPosition{0};
-    bool isEmpty{false};
-
-    do
-    {
-        /// @todo iox-#1695 read before write since the writer increments the aba counter!!!
-        /// @todo iox-#1695 write doc with example!!!
-        currentReadPosition = m_readPosition.load(std::memory_order_acquire);
-        uint64_t currentWritePosition = m_writePosition.load(std::memory_order_acquire);
-
-        isEmpty = (currentWritePosition == currentReadPosition);
-        // we need compare without exchange
-    } while (!(currentReadPosition == m_readPosition.load(std::memory_order_acquire)));
-
-    return isEmpty;
+    auto [readPosition, writePosition] = getReadWritePositions();
+    return readPosition == writePosition;
 }
 
 template <class ValueType, uint64_t CapacityValue>
 inline bool SpscSofi<ValueType, CapacityValue>::pop(ValueType& valueOut) noexcept
 {
-    return popIf(valueOut, [](ValueType) { return true; });
-}
-
-template <class ValueType, uint64_t CapacityValue>
-template <typename Verificator_T>
-inline bool SpscSofi<ValueType, CapacityValue>::popIf(ValueType& valueOut, const Verificator_T& verificator) noexcept
-{
+    // Memory synchronization is not needed but we need to prevent operation reordering to avoid the following scenario
+    // where the CPU reorder the load of m_readPosition and m_writePosition:
+    // 0. Initial situation (the queue is full)
+    // |----|--B--|--C--|
+    // ^     ^
+    // w=3  r=1
+    // 1. The consumer thread loads m_writePosition => 3
+    // |----|--B--|--C--|
+    // ^     ^
+    // w=3  r=1
+    // 2. The producer thread pushes two times
+    // |--D--|--E--|-----|
+    // ^           ^
+    // r=3        w=5
+    // 3. The consumer thread loads m_readPosition => 3. The pop method returns false
+    // => Whereas the queue was full, pop returned false giving the impression that the queue if empty
```

Comment:
Can you move the explanation to
```cpp
        if (currentReadPosition == m_writePosition.load(std::memory_order_acquire))
        {
            nextReadPosition = currentReadPosition;
            popWasSuccessful = false;
        }
```
below. It is not related to the memory order but the reason on cannot just call `return false;` but needs to continue with the loop.

The `acquire` just corresponds to the `m_readPosition.compare_exchange_strong` store operation in push.

Reply:
I don't understand. You're saying that without any operation reordering and by returning false, we can have the following scenario:

```
0. Initial situation (the queue is full)
|----|--B--|--C--|
^    ^
w=3 r=1
1. The consumer thread loads m_writePosition => 3
|----|--B--|--C--|
^     ^
w=3  r=1
2. The producer thread pushes two times
|--D--|--E--|-----|
^           ^
r=3        w=5
3. The consumer thread loads m_readPosition == m_writePosition so the pop method returns false
=> Whereas the queue was full, pop returned false giving the impression that the queue if empty
```

Reply:
You are right. I confused it with a previous version of the patch where the load of `m_readPosition` was relaxed. With `acquire` this cannot happen. I think it should now even be possible to have an early return. Unfortunately we do not have the full history of the SpscSofi but I think we once had a slightly different algorithm and kept `popWasSuccessful` after the refactoring.

The comment should be moved back where it was and an early return can be introduced.

Reply:
Changed

Reply:
Hehe, now you get some warnings because `nextReadPosition` is set but not used anymore and the clang-tidy also does not like to have an else branch after an early return :)

Reply:
Changed